compiler-lib/src/semantics.rs

```diff
@@ -124,6 +127,47 @@ impl<'a, 'f, 'cx> ClassesAndMembersVisitor<'a, 'f, 'cx> {
                         }
                         self.visit_static_block(&NodeKind::from(block), &params[0].name);
                     }
+
+                    match &member.kind {
+                        ast::ClassMemberKind::Method(ty, _, block)
+                            if ty.basic != ast::BasicType::Void =>
+                        {
+                            fn always_returns<'t>(stmt: &Spanned<'t, ast::Stmt<'t>>) -> bool {
+                                match &stmt.data {
+                                    // An if-else stmt always return iff both arms always return
+                                    ast::Stmt::If(_, then_arm, else_arm) => {
+                                        let then_arm_always_returns = always_returns(&*then_arm);
+                                        let else_arm_always_returns = else_arm
+                                            .as_ref()
+                                            .map_or(true, |else_arm| always_returns(&*else_arm));
+
+                                        then_arm_always_returns && else_arm_always_returns
+                                    }
+
+                                    // A block always returns iff any of the statements always
+                                    // returns
+                                    ast::Stmt::Block(block) => {
+                                        block.statements.iter().any(always_returns)
+                                    }
+
+                                    // A return stmt always returns
+                                    ast::Stmt::Return(_) => true,
+
+                                    // All other stmts do not always return
+                                    _ => false,
+                                }
+                            }
+
+                            if !block.statements.iter().any(always_returns) {
+                                self.context.diagnostics.error(&Spanned {
+                                    span: block.span.clone(),
```

Comment:
The span for the method declaration does not seam to be directly available. But it can probably be reconstructed out of `Method(
        Spanned<'t, Type<'t>>,
        Spanned<'t, ParameterList<'t>>,
        Spanned<'t, Block<'t>>,
),` by combining the spans of Type and ParameterList?

Reply:
(This specific suggestion is not blocking.)

Reply:
If you want to keep the block in the span, at least extend it to the signature by combining the first member of method (return type) with the block.span

Reply:
https://github.com/comprakt/comprakt/pull/98/commits/5a827df2904630928972852920d75cd368462eaf adds this, but it's pretty ugly since we currently do not have a link from `ClassMember` to its span.
The alterantive to the HashMap proposed in above commit to add a copy of a `ClassMember`'s full span (`ClassDecl.members[i].span`)  to the ClassMember.
Or we could change the visitor to also pass spans for each child, e.g. as a second closure argument. However, I think we only really _need_ this for `ClassMember`, both in first and second pass (second pass for mismatched return types, etc).